flang/include/flang/Lower/PFTBuilder.h

```diff
@@ -90,6 +90,25 @@ using MutableReferenceVariant = ReferenceVariantBase<false, A...>;
 using PftNode = MutableReferenceVariant<Program, ModuleLikeUnit,
                                         FunctionLikeUnit, Evaluation>;
 
+/// Class to be inherited by all PFT nodes that have parents.
+/// It provides the ability to traverse the PFT tree upwards.
+struct ChildNode {
```

Comment:
This name seems backwards as "ChildNode" is being used to build a parent relation.

Is it really needed? Playing Devil's advocate: in the parse tree, inheritance like this is rejected in favor of denotational convergence. For example, every node in the PFT could have a member named `parent` when applicable (which I think is already the case), and a template function can be written to access `that.parent` when `that` is any type `a` with such a member. (i.e., like `.u`, `.t`, and `.v` members in the raft of parse tree classes.)

```c++
template <typename ParentType, typename A>
  ParentType *getAncestor(const A &that) {
    if (auto *seekedParent = that.parent.getIf<ParentType>())
      return seekedParent;
    return parent.visit(
        common::visitors{[](Program &p) -> ParentType * { return nullptr; },
                         [](auto &p) -> ParentType * {
                           return template getAncestor<ParentType>(p);
                         }});
  }
```

Reply:
> Is it really needed?

No, I thought it was nicer to use a bit of class hierarchy to confer the property of being a child node. But your suggestion is less intrusive,. I went for it, thanks.